### ISSUE TITLE
[DataGrid] "Cannot read property 'id' of undefined" when rows update

### ISSUE THREAD

**Issue by PeppermintNDaisies** [mui-org/material-ui-x#575]
I submitted a ticket to zendesk first and was asked to create an issue here. I was also told this is similar to issue #571  

- [x] The issue is present in the latest release.
- [x] I have searched the [issues](https://github.com/mui-org/material-ui-x/issues) of this repository and believe that this is not a duplicate.

## Current Behavior 😯

I'm working with XGrid and since the Filter feature is not yet available, I'm trying to implement it manually, but I keep getting this error (see screenshot and screen recording bellow) and can't figure out what I'm doing wrong. Essentially what I do is re render all the rows (like in the docs https://material-ui.com/components/data-grid/rows/#the-rows-prop).

![Screen Shot 2020-11-12 at 3 25 07 PM](https://user-images.githubusercontent.com/27710492/98998716-94b63c80-2504-11eb-8a2e-d0ce07dc42ca.png)

![Hnet com-image](https://user-images.githubusercontent.com/27710492/98999592-0b076e80-2506-11eb-862b-9b9975eea0d3.gif)

## Expected Behavior 🤔

Updating rows and/or columns should work as expected without error.

## Steps to Reproduce 🕹

Tried to reproduce but it's kind of hard to reproduce all the condition since all of the data is brought from a backend, but this [CodeSandbox](https://codesandbox.io/s/youthful-elgamal-gzzl4?file=/src/App.js) link is more or less how I manage the data.

Steps:

1. The rows array get 're calculated' causing the re render. The rows are fewer than before (trying to actually implement Filtering feature manually)
2. Error occurs

## Context 🔦

Basically implement Filtering feature manually.

## Your Environment 🌎

| Tech          | Version |
| ------------- | ------- |
| Material-UI X | v4.0.0-alpha.8|
| React         |    ^16.13.1     |
| Browser       |    Chrome     |
| TypeScript    |     yes   |

**Comment by oliviertassinari** [comment]
I can't reproduce the exception on the codesandbox. We could wait for the resolution of #571 to see if it solves your case but without a reproduction, we operate in the dark. You are depending on the chance for the outcome.

**Comment by orhels** [comment]
Been having the same issues. When updating the DataGrid from the backend with new or updated columns/row we get the `Cannot read property 'id' of undefined` error.
Updated the sandbox @PeppermintNDaisies supplied to reproduce the error: https://codesandbox.io/s/sharp-hawking-20q80. Just press the "Reproduce bug"-button two times and the grid crashes.

**Comment by PeppermintNDaisies** [comment]
@orhels Thank you so much ✨ ! I hope with this reproduction the problem can be looked into and fixed

**Comment by BrentFarese** [comment]
I submitted #571.  Just FYI for everyone here, I bumped down a version to `4.0.0-alpha.8` and I am not experiencing the issue anymore.  Still have more testing to do, but at least that might narrow down the issue/bug to a specific release @oliviertassinari.  Thanks!

**Comment by matandro** [comment]
> I submitted #571. Just FYI for everyone here, I bumped down a version to `4.0.0-alpha.8` and I am not experiencing the issue anymore. Still have more testing to do, but at least that might narrow down the issue/bug to a specific release @oliviertassinari. Thanks!

Had the same issue and downgraded to alpha8 and my filter now works.

**Comment by SvenFaerber** [comment]
While updating the sortedRows should fix the problem for this specific issue, I believe a related issue (which brought me here) exists for the selectedRows, which are not updated when rows change either.

**Comment by oliviertassinari** [comment]
The fix can be tried with:

```json
"@material-ui/data-grid": "https://pkg.csb.dev/mui-org/material-ui-x/commit/02423225/@material-ui/data-grid",
"@material-ui/x-grid": "https://pkg.csb.dev/mui-org/material-ui-x/commit/02423225/@material-ui/x-grid",
```

**Comment by oliviertassinari** [comment]
So far, we have got one valid reproduction from @orhels: https://codesandbox.io/s/sharp-hawking-20q80. which is solved: https://codesandbox.io/s/nice-frog-togvz?file=/package.json. 

@SvenFaerber You would need to provide a new reproduction in a new issue.

**Comment by orhels** [comment]
@oliviertassinari testet it in my project and it worked flawlessly. Thanks to you and @dtassone for the hard work 👏🏼

**Comment by arhirani** [comment]
I faced the same issue, got resolved by deprecating the version to 4.0.0-alpha.8. This thread helped a lot. Thanks everyone.
